README.md
# ReCalc v1.3.3

Calculadora con funciones simples.

## Dependencias

- node > 18

## Comandos

```bash
# iniciar/limpiar la db
npm run seed

# iniciar calculadora
npm run start

# correr lost test
npm run test
```
## Integrantes
- Victoria Heredia
- Franco Tastaca


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.3] 21-06-2023

### Refactored

- Test deleteAllHistory
- Orden de la persistencia

### Fixed

- Conexión de las funciones con la API
- Problema al correr tests lint
- Bug “undefined” en espacio vacío

### Added

- Test propio para el nuevo atributo error
- Tests e2e de las funciones
- Implementación del botón C

### Changed

- Interfaz básica de la calculadora

## [1.3.2] 08-06-2023

### Fixed

- Path donde corre eslint

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.1] 05-06-2023

### Added

- Dependencia cross-env para correr tests e2e

### Fixed

- Problema al correr tests e2e en Windows

## [1.3.0] 05-06-2023

### Added

- Interfaz básica de la calculadora
- ESLint para tests estáticos
- Playwright para tests e2e

## [1.2.0] 15-05-2023

### Added

- Sequelize para el manejo de base de datos
- Tests de integración sobre API y modelo
- Integración continua

## [1.1.0] 20-04-2023

### Added

- Endpoint para la función de resta
- Dependencias para realizar tests
- Tests para la función de resta

## [1.0.0] 17-04-2023

### Added

- Interface CLI para realizar cálculos en forma interactiva
- Base de la API
- Funciones core para realizar suma, resta, multiplicación, división y potencia

[unreleased]: https://github.com/FrancoTastaca/recalc/compare/v1.3.3...HEAD
[1.3.3]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.3.3
[1.3.2]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.3.2
[1.3.1]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.3.1
[1.3.0]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.3.0
[1.2.0]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.2.0
[1.1.0]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.1.0
[1.0.0]: https://github.com/FrancoTastaca/recalc/releases/tag/v1.0.0


test/ui.test.js
import { test, expect } from '@playwright/test';
import { seed } from '../src/seed.js';
import { Operation, History } from '../src/models.js'

test.describe('test', () => {
  test.describe.configure({ mode: 'serial' });

  test.beforeEach(async () => {
    await seed();
  })

  test('Deberia tener como titulo de pagina recalc', async ({ page }) => {
    await page.goto('./');

    // Expect a title "to contain" a substring.
    await expect(page).toHaveTitle(/recalc/i);
  });

  test('Deberia poder realizar una resta', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '7' }).click()
    await page.getByRole('button', { name: '9' }).click()
    await page.getByRole('button', { name: '-' }).click()
    await page.getByRole('button', { name: '9' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sub/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(70);

    await expect(page.getByTestId('display')).toHaveValue(/70/)

    const operation = await Operation.findOne({
      where: {
        name: "SUB"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(79)
    expect(historyEntry.secondArg).toEqual(9)
    expect(historyEntry.result).toEqual(70)
  });

  test('Deberia poder realizar una multiplicacion', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '2', exact: true }).click()
    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '*' }).click()
    await page.getByRole('button', { name: '2', exact: true }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/mul/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(50);

    await expect(page.getByTestId('display')).toHaveValue(/50/)

    const operation = await Operation.findOne({
      where: {
        name: "MUL"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(25)
    expect(historyEntry.secondArg).toEqual(2)
    expect(historyEntry.result).toEqual(50)
  });

  test('Deberia poder realizar una suma', async ({ page }) => {
    await page.goto('./');


    await page.getByRole('button', { name: '1' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '+' }).click()
    await page.getByRole('button', { name: '3' }).click()


    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/add/')),
      page.getByRole('button', { name: '=' }).click()
    ]);


    const { result } = await response.json();
    expect(result).toBe(13);


    await expect(page.getByTestId('display')).toHaveValue(/13/)


    const operation = await Operation.findOne({
      where: {
        name: "ADD"
      }
    });


    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })


    expect(historyEntry.firstArg).toEqual(10)
    expect(historyEntry.secondArg).toEqual(3)
    expect(historyEntry.result).toEqual(13)
  });

  test('Deberia poder calcular la potencia cuadrada de un número', async ({page}) => {
    await page.goto('./');


    await page.getByRole('button', { name: '1' }).click();
    await page.getByRole('button', { name: '0' }).click();
    await page.getByRole('button', { name: '^2' }).click();

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/pow/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(100);

    await expect(page.getByTestId('display')).toHaveValue(/100/)

    const operation = await Operation.findOne({
      where: {
        name: "POW"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(10)
    expect(historyEntry.result).toEqual(100)
  });
  test('Deberia verificar que la calculadora tire error si el resultado es mayor que 100.000', async ({page}) => {
    await page.goto('./');


    await page.getByRole('button', { name: '1' }).click();
    await page.getByRole('button', { name: '2', exact: true }).click()
    await page.getByRole('button', { name: '3' }).click();
    await page.getByRole('button', { name: '4' }).click();
    await page.getByRole('button', { name: '5' }).click();
    await page.getByRole('button', { name: '6' }).click();
    await page.getByRole('button', { name: '7' }).click();
    await page.getByRole('button', { name: '^2' }).click();
    await page.getByRole('button', { name: '=' }).click();
    await expect(page.getByTestId('display')).toHaveValue(/Error:n° muy grande, pruebe con otro/)
  });
  test('Deberia poder realizar una división', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '1' }).click()
    await page.getByRole('button', { name: '2', exact: true }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '6' }).click()
    await page.getByRole('button', { name: '0' }).click()

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/div/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(0.2);

    await expect(page.getByTestId('display')).toHaveValue(/0.2/)

    const operation = await Operation.findOne({
      where: {
        name: "DIV"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(12)
    expect(historyEntry.secondArg).toEqual(60)
    expect(historyEntry.result).toEqual(0.2)
  });
  test('Deberia mostrar un mensaje de error en el display si el divisor es 0', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '2', exact: true }).click()
    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '/' }).click()
    await page.getByRole('button', { name: '0' }).click()
    await page.getByRole('button', { name: '=' }).click()
    await expect(page.getByTestId('display')).toHaveValue(/¡Error! No se puede dividir por 0/);
  });
  test('Deberia limpiar el display al precionar el boton C', async ({ page }) => {
    await page.goto('./'); 
    await page.getByRole('button', { name: '5' }).click()
    await page.getByRole('button', { name: '*' }).click() 
    await page.getByRole('button', { name: '^2' }).click()
    await page.getByRole('button', { name: 'c'}).click(); 
    await expect(page.getByTestId('display')).toHaveValue('') 
  })
  test('No deberia hacer nada al hacer clic en un lugar vacio', async ({ page }) => {
    await page.goto('./');
  
    await page.getByRole('button', { name: '7' }).click();
    await page.getByRole('button', { name: '-' }).click();
    await page.evaluate(() => {
      const emptyArea = document.querySelector('.empty-area'); // Reemplaza '.empty-area' con el selector adecuado para el área vacía
      if (emptyArea) {
        emptyArea.click();
      }
    });
    await expect(page.getByTestId('display')).toHaveValue('7-');
  });
  test('Deberia poder calcular la raíz cuadrada de un número positivo', async ({page}) => {
    await page.goto('./');


    await page.getByRole('button', { name: '1' }).click();
    await page.getByRole('button', { name: '6' }).click();
    await page.getByRole('button', { name: '√' }).click();

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/sqrt/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(4);

    await expect(page.getByTestId('display')).toHaveValue(/4/)

    const operation = await Operation.findOne({
      where: {
        name: "SQRT"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })
    
    expect(historyEntry.firstArg).toEqual(16)
    expect(historyEntry.result).toEqual(4)
  });
  test('Deberia mostrar un mensaje de error en el display si el número es negativo para sqrt', async ({ page }) => {
    await page.goto('./');

    await page.getByRole('button', { name: '-', exact: true }).click()
    await page.getByRole('button', { name: '8' }).click()
    await page.getByRole('button', { name: '√' }).click()
    await page.getByRole('button', { name: '=' }).click()
    await expect(page.getByTestId('display')).toHaveValue(/¡Error! El n° debe ser positivo/);
  });
  test('Deberia poder convertir decimal a binario', async ({page}) => {
    await page.goto('./');
    
    await page.getByRole('button', { name: '5' }).click();
    await page.getByRole('button', { name: 'dtb' }).click();

    const [response] = await Promise.all([
      page.waitForResponse((r) => r.url().includes('/api/v1/dtb/')),
      page.getByRole('button', { name: '=' }).click()
    ]);

    const { result } = await response.json();
    expect(result).toBe(101);

    await expect(page.getByTestId('display')).toHaveValue(/101/)

    const operation = await Operation.findOne({
      where: {
        name: "DTB"
      }
    });

    const historyEntry = await History.findOne({
      where: { OperationId: operation.id }
    })

    expect(historyEntry.firstArg).toEqual(5)
    expect(historyEntry.result).toEqual(101)
  });  
})

test/api.test.js
const request = require('supertest');
const api = require('../src/api.js');
const { seed } = require('../src/seed.js')

beforeEach(async () => {
    await seed()
})

describe("API substract", () => {
    test("Deberia responder con un 200 ok", async () => {
        const app = await api.build()

        return request(app).get('/api/v1/sub/2/1')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(1);
            })
    })
})

describe("API add", () => {
    test("El segundo parámetro es negativo, por lo que el resultado debe ser menor al primer parametro por ser una suma y el endpoint debe devolver un status 200", async () => {
        const app = await api.build();

        return request(app).get('/api/v1/add/10/-5')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeLessThan(10); 
            })
    });
    test("El resultado de la suma de 0.1 y 0.2 debe ser 0.3", async () => {
        const app = await api.build();

        return request(app).get('/api/v1/add/0.1/0.2')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeCloseTo(0.3); 
            })
    });
});

describe("API Mul", () => {
    test("Ambos parametros son decimales, por lo que el resultado debe ser con decimales y el thenpoint debe devolver un status 200", async () => {
        const app = await api.build();
        return request(app).get('/api/v1/mul/2.124/4.757')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(Number.isInteger(res.body.result)).toBe(false) 

            })
    });
});

describe("API Div", () => {
    test("Si el segundo parámetro es 0, el thenpoint debe devolver un mensaje de error junto a un status 400", async () => {
        const app = await api.build();
        return request(app).get('/api/v1/div/8/0')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual('¡ERROR! No se puede dividir por 0 ') 
            })
    });
});

describe.skip("API Pow", () => {
    test("Si el parámetro no es un número, el thenpoint debe devolver un error junto a un status 400", async () => {
        const app = await api.build();
        return request(app).get('/api/v1/pow/t')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual('El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido para la potencia cuadrada') 
            })
    });
});

//Ocurre el mismo problema que con POW API... pero cuando estan ambas sin comentar - se traba primero en API POW y pasa este. Pero cuando comente el anterior y quedo este solo, dispara el mismo error de que no encuentra la operacion.
describe.skip("API Sqrt", () => {
    test("Debería calcular correctamente la raíz cuadrada de 16 y devolver un resultado de 4", async () => {
        const app = await api.build();

        return request(app).get('/api/v1/sqrt/16')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(4)
            })
    })
    test("Debería devolver un código de estado 400 al calcular la raíz cuadrada de un número no válido", async () => {
        const app = await api.build();
    
        return request(app).get('/api/v1/sqrt/abc')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual('El parámetro ingresado no es un número válido o es negativo. Por favor, asegúrese de que sea un parámetro válido y no negativo para la raiz cuadrada')
            })
    })
    test("Debería calcular correctamente la raíz cuadrada de 2 y devolver un resultado aproximadamente 1.414", async () => {
        const app = await api.build();
    
        return request(app).get('/api/v1/sqrt/2')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeCloseTo(1.414);
            })
    })
})

//Ocurre el mismo problema que con POW API
describe.skip("API Dtb", () => {
   test("Debería responder con un 200 ok", async () => {
       const app = await api.build();

       return request(app).get('/api/v1/dtb/0.25')
           .expect(200)
           .expect('Content-Type', "application/json; charset=utf-8")
           .then((res) => {
               expect(res.body.result).toBeCloseTo(0.01);
           })
   })

   test("Debería responder con un estatus 200 ok", async () => {
        const app = await api.build();

        return request(app).get('/api/v1/dtb/8')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toEqual(1000);
            })
    })
    
    test("Debería devolver un código de estado 400 al intentar convertir un parámetro que no es numérico", async () => {
        const app = await api.build();
    
        return request(app).get('/api/v1/dtb/t')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual('El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido.')
            })
   })
})

describe.skip("API historyById", () => {
    test("Si el parámetro ingresado no es un valor numérico, el thenpoint debe devolver un mensaje de error junto a un status 400", async () => {
        const app = await api.build();
        return request(app).get('/api/v1/historyById/id')
            .expect(400)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.error).toEqual('El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido.') 
            })
    });

    test("Debería devolver un 200 ok", async () => {
        const app = await api.build();
        return request(app).get('/api/v1/historyById/2')
            .expect(200)
            .expect('Content-Type', "application/json; charset=utf-8")
            .then((res) => {
                expect(res.body.result).toBeDefined(); 
            })
    });

});



test/core.test.js
import core from '../src/core.js'
import { validacionParametrosNum} from '../src/controllers.js'

describe('Subtract', () => {
    test('Debería 2 - 2 = 0', () => {
        expect(core.sub(2, 2)).toBe(0); 
    })

    test('Debería 6 - 4 = 2', () => {
        expect(core.sub(6, 4)).toBe(2); 
    })
})

describe('Add', () => {
    test('Debería 240 + 16 = 256', () => {
        expect(core.add(240, 16)).toBe(256); 
    })

    test('Debería -36 + 55 = 19', () => {
        expect(core.add(-36, 55)).toBe(19); 
    })
    
    test('Debería sumar correctamente si ambos parámetros son números', () => {
        const req = {
            params: { a: '2', b: '2' } //Simulamos una peticion de HTTP de los parametros a y b 
        };
        /*jest.fn() es una función proporcionada por Jest que se utiliza para crear una "función ficticia" que se puede usar como sustituto de una función real en las pruebas. En este caso, se crea una función "falsa next() que se pasará como tercer parámetro a validacionParametrosNum. Al llamar a la función next dentro de validacionParametrosNum, se puede comprobar que se ha llamado correctamente con expect(next).toHaveBeenCalled(). */
        const next = jest.fn(); 
        validacionParametrosNum(req, null, next);
        const result = core.add(req.validParams.a, req.validParams.b);
        expect(result).toBe(4); 
        expect(next).toHaveBeenCalled();
    })
    /*test('Debería lanzar un error si alguno de los parámetros no es numérico', () => {
        const req = {
            params: { a: '1', b: 'no es un número' }
        };
        const next = jest.fn();
        const res = {
          status: jest.fn().mockReturnThis(),
          json: jest.fn(),
        }
        validacionParametrosNum(req, res, next);
        expect(res.status).toHaveBeenCalledWith(400);
        expect(res.json).toHaveBeenCalledWith({
          error:
            'Uno de los parámetros no es un número. Por favor, asegúrese de que ambos parámetros sean válidos',
        });
      }); */
})


describe('Divide', () => {
    test('Debería hacer una división exacta 21 / 3  = 7', () => {
        expect(core.div(21, 3)).toBe(7);
        expect(core.mul(3, 7)).toBe(21);
        expect(21 % 3).toBe(0); 
    })

    test('Debería hacer una división inexacta 67 / 8  = 8.375', () => {
        expect(core.div(67, 8)).toBe(8.375);
        //Comprobación con decimal
        expect(core.mul(8.375, 8)).toBe(67);
        //Para la siguiente comprobación se tuvo en cuenta el número entero, no el decimal
        expect(core.add((core.mul(8, 8)), (67 % 8))).toBe(67);
        expect(67 % 8).not.toBe(0);
        expect(67 % 8).toBeLessThan(8); 
    })
    
    test('Debería hacer una división con divisor mayor al dividendo 6 / 24  = 0.25', () => {
        expect(core.div(6, 24)).toBe(0.25); 
    })

    test('Debería dividir con números negativos', () => {
        expect(core.div(-5, 2)).toBeLessThan(0);
        expect(core.div(12, -30)).toBeLessThan(0);
        expect(core.div(-24, -12)).toBe(2); 
    })
})

describe('Pow', () => {
    test('Debería 9² = 81', () => {
        expect(core.pow(9)).toBe(81);
        
    })

    test('Debería dar cero toda potencia con base igual a 0', () => {
        expect(core.pow(0)).toBe(0); 
    })

    test('Debería dar uno toda potencia con base igual a 1', () => {
        expect(core.pow(1)).toBe(1); 
    })
})

describe('Multiply', () => {
  test('Debería multiplicar dos números positivos correctamente', () => {
    expect(core.mul(2, 2)).toBe(4);
  })

  // Tratamos algunas de las propiedades de multiplicacion 
  test('Debería cumplir con la propiedad conmutativa de la multiplicación', () => {
    expect(core.mul(2, 3)).toBe(core.mul(3, 2));
  })

  test('Debería cumplir con la propiedad de identidad de la multiplicación', () => {
    expect(core.mul(4, 1)).toBe(4);
    expect(core.mul(0, 1)).toBe(0);
    expect(core.mul(-2, 1)).toBeLessThan(0);
  })
})

describe('Sub-Opcional', () => {
    test('Debería dar un número negativo una resta con el segundo parámetro mayor al primero', () => {
        expect(core.sub(5, 30)).toBeLessThan(0);
        expect(core.sub(90, 200)).toBeLessThan(0);
        expect(core.sub(-43, -20)).toBeLessThan(0);
        expect(core.sub(-18, 5)).toBeLessThan(0);
    })
})

describe('Mul-Opcional', () => {
    test('Debería multiplicar un número positivo y un número negativo correctamente', () => {
        expect(core.mul(2, -2)).toBeLessThan(0);
        expect(core.mul(-2, 2)).toBeLessThan(0);
      })
    
      test('Debería multiplicar dos números negativos correctamente', () => {
        expect(core.mul(-2, -2)).toBe(4);
      })
})

describe('Pow-Opcional', () => {
    test('Debería dar un número positivo toda potencia con base negativa ', () => {
        expect(core.pow(-55)).toBe(3025);
        expect(core.pow(-24)).toBeGreaterThan(0);
      })
})
describe('Div-Opcional', () => {
    test('Debería devolver un error al dividir por 0', () => {
        expect(core.div(45,0)).toMatch('¡ERROR! No se puede dividir por 0 ');
    })
})
describe('Sqrt-Opcional', () => {
    test('Debería calcular correctamente la raiz cuadrada de 16', () => {
      expect(core.sqrt(16)).toBe(4);
    });
    test('Debería aproximar correctamente la raíz cuadrada de un número decimal', () => {
        expect(core.sqrt(2)).toBeCloseTo(1.414);
      });
})

describe('Dtb-Opcional', () => {
    test('El número decimal 20 debería ser 10100 en binario', () => {
        expect(core.dtb(20)).toBe(10100);
    });
    test('El número decimal 6.7 debería ser cercano a 110.101100110011 en binario', () => {
        expect(core.dtb(6.7)).toBeCloseTo(110.101100110011);
    });
})


src/cli.js
import { createInterface } from "node:readline/promises";
import core from "./core.js"

const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
});

const AVAILABLE_FNS = [...Object.keys(core), 'exit'].join(', ')

async function loop() {
    const fnName = await readline.question(`Ingrese función (${AVAILABLE_FNS}): `)
    const fn=core[fnName];

    if (fnName === "exit") {
        console.log("👋👋👋");
        return readline.close();
    }
    if (!fn) {
        console.log(`La función "${fnName}" es incorrecta, por favor vuelva intentar`);
    }
    /*La función const fn = core[fnName]; toma el valor de core[fnName], es decir, toma el valor de la propiedad con nombre igual al valor de la variable fnName dentro del objeto core. Si fnName es 'add', fn tomará el valor de la función add dentro de core.
    Luego, el if (!fn) verifica si fn es falso, lo que sucederá si core[fnName] no existe o es undefined. Si es así, se imprimirá un mensaje de error en la consola con el nombre de la función incorrecta, y se llamará a la función loop() para volver a pedir al usuario que ingrese una función válida. */ 

    else if (fnName === "pow"){
            const powNum = await readline.question("Ingrese el número a elevar al cuadrado: ");
            //Comprueba si se está ingresando un valor numérico, de lo contrario, se solicita que debe ingresar un número
            if (! isNaN(powNum)){
                //Se trata el bug de solicitar mas parametros de los necesarios para el pow cuadrado
                console.log(`El resultado del cuadrado es: ${core.pow(Number(powNum))}`);
            } else {
                console.log('Debe ingresar un número');
            }      
        }
    else {
        const firstNum = await readline.question("Ingrese el primer número: ")
        if (! isNaN(firstNum)){
            const secondNum = await readline.question("Ingrese el segundo número: ")
            if (! isNaN(secondNum)){
                switch(fnName) {
                    case "add":
                        console.log(`El resultado de la suma es: ${core.add(Number(firstNum), Number(secondNum))}`);
                        break;
                    case "sub":
                        console.log(`El resultado de la resta es: ${core.sub(Number(firstNum), Number(secondNum))}`);
                        break;
                    case "mul":
                        console.log(`El resultado de la multiplicacion es: ${core.mul(Number(firstNum), Number(secondNum))}`);
                        break;    
                    case "div":
                    console.log(`El resultado de la division es: ${core.div(Number(firstNum), Number(secondNum))}`);
                }
            } else {
                console.log('Debe ingresar un número');
            }
        } else {
            console.log('Debe ingresar un número');
        }
    }
    loop();
    }
    
loop();

src/core.js
function add(a, b) {
    let result = a + b;
    return result;
}

function subtract(a, b) {
    let result = a - b;
    return result;
}

function multiply(a, b) {
    let result = a * b;
    return result;
}

function divide(a, b) {
    let result = (b !== 0) ? (a/b) : "¡ERROR! No se puede dividir por 0 "; //Otra forma mas 
    // La logica de sintaxis es la siguiente: "condicion ? valorSiVerdadero : valorSiFalso;" 
    return result;
}

function pow(a) { 
    let result= Math.pow(a,2);
    return result;
}
function sqrt(a) {
    let result= Math.sqrt(a); 
    return result;
}

function decimalToBinary(a) {
    let result = parseFloat((parseFloat(a, 10)).toString(2));
    return result;
}


export default {
    add: add,
    sub: subtract,
    mul: multiply,
    div: divide,
    pow: pow,
    sqrt:sqrt,
    dtb: decimalToBinary
}


src/controllers.js
import express from 'express';
import core from './core.js';

import { createHistoryEntry,getAllHistory,deleteAllHistory,historyById } from './models.js'

const router = express.Router();

function guardarValoresConError(req, error) {

    let operationName = ''
    const routePath = req.route.path; 
    //req.route.path contiene la ruta completa,
    // Con el startWith podemos hacer foco en como inicia la ruta y nos desligamos de la dependencia de los valores de entrada, ya que sabemos que no son numeros
    if (routePath.startsWith('/sub')) {
      operationName = 'SUB'
    } else if (routePath.startsWith('/add')) {
      operationName = 'ADD'
    } else if (routePath.startsWith('/div')) {
      operationName = 'DIV'
    } else if (routePath.startsWith('/mul')) {
      operationName = 'MUL'
    }
    createHistoryEntry({ firstArg: null, secondArg: null, operationName, error: error.message })
  }
function validacionParametrosNum(req, res, next) {
    try {
        const params = req.params;
        const a = Number(params.a);
        const b = Number(params.b);
        if (isNaN(a) || isNaN(b)) {
            throw new Error('Uno de los parámetros no es un número. Por favor, asegúrese de que ambos parámetros sean válidos');
        }
        req.validParams = { a, b };
        next();
    } catch (error) {
      guardarValoresConError(req, error);
      res.status(400).json({ error: error.message });
    }
}

router.get("/sub/:a/:b", validacionParametrosNum, async function (req, res) {
    const { a, b } = req.validParams;
    const result = core.sub(a, b);
    await createHistoryEntry({ firstArg: a, secondArg: b, operationName: "SUB", result})
    res.send({ result });
});

router.get("/add/:a/:b", validacionParametrosNum, async function (req, res) {
        const { a, b } = req.validParams;
        const result = core.add(a, b);
        await createHistoryEntry({ firstArg: a, secondArg: b, operationName: "ADD", result})
        res.send({ result });
});

router.get("/div/:a/:b", validacionParametrosNum, async function (req, res) {
    const { a, b } = req.validParams;
    if ( b === 0) {
        const errorMsg = '¡ERROR! No se puede dividir por 0 '
        await createHistoryEntry({ firstArg: a, secondArg: b, operationName: "DIV", error: errorMsg})
        res.status(400).send({ error: errorMsg})
    }else{
        const result = core.div(a, b);
        await createHistoryEntry({ firstArg: a, secondArg: b, operationName: "DIV", result,})
        res.send({ result });
    }
});

router.get("/pow/:a", async function (req, res) {
    const a = Number(req.params.a);
    if (isNaN(a)) {
        const errorMsg ='El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido para la potencia cuadrada'
        await createHistoryEntry({ firstArg: null, operationName: "POW" , error:errorMsg })
        res.status(400).send({ error: errorMsg})
    } else {
        const result = core.pow(a);
        await createHistoryEntry({ firstArg: a, operationName: "POW", result});
        res.send({ result });
    }
});

router.get("/mul/:a/:b", validacionParametrosNum, async function (req, res) {
    const { a, b } = req.validParams;
    const result = core.mul(a, b);
    await createHistoryEntry({ firstArg: a, secondArg: b, operationName: "MUL", result});
    res.send({ result });
});

router.get("/sqrt/:a", async function (req, res) {
    const a = Number(req.params.a);
    if (isNaN(a)|| a < 0) {
        const errorMsg = 'El parámetro ingresado no es un número válido o es negativo. Por favor, asegúrese de que sea un parámetro válido y no negativo para la raiz cuadrada'
        await createHistoryEntry({ firstArg: null, operationName: "SQRT", error: errorMsg});
        res.status(400).send({ error:errorMsg});
    } else {
        const result = core.sqrt(a);
        await createHistoryEntry({ firstArg: a, operationName: "SQRT", result});
        res.send({ result });
    }
});

router.get("/dtb/:a", async function (req, res) {
    const a = Number(req.params.a);
    if (isNaN(a)) {
        const errorMsg = 'El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido.'
        await createHistoryEntry({ firstArg: null, operationName: "DTB", error: errorMsg});
        res.status(400).send({ error:errorMsg});
    } else {
        const result = core.dtb(a);
        await createHistoryEntry({ firstArg: a, operationName: "DTB", result});
        res.send({ result });
    }
});

router.get("/historyById/:id", async function (req, res) {
    const id = Number(req.params.id);
    if (isNaN(id)) {
        res.status(400).send({error: 'El parámetro ingresado no es un número. Por favor, asegúrese de que sea un parámetro válido.'})
    }else{
        const result = await historyById(id);
        res.send({ result });
    }
});

router.get("/history", async function (req, res) {
    const history = await getAllHistory()
    // Mapeamos los datos para formatear y estructurar mejor la legibilidad de los datos
    const formattedHistory = history.map(({ id, Operation, firstArg, secondArg, result, error}) => ({
      id,
      operation: Operation.name,
      arguments: {
        first: firstArg,
        second: secondArg
      },
      result,
      error,
    }))
    //Generamos una respuesta en HTML que muestra la cadena JSON preformateada dentro de la etiqueta <pre>. Al hacerlo, logramos una visualización más legible y de facil lectura
    res.send(`
     <pre>${JSON.stringify({ history: formattedHistory}, null, 2)}</pre> 
    `)
    
});

// Queda pendiente buscar la forma de que la ruta "/deleteHistory" utilice el método DELETE en lugar de GET, ya que es lo más apropiado para responder a una solicitud de eliminación del historial. Sin embargo, por el momento, se mantiene la ruta de esta forma para poder utilizar esta funcionalidad.

router.get("/deleteHistory",async function(req,res){
    await deleteAllHistory()
    res.status(200).send({ message: "Historial eliminado con exito" })
     
});

export { validacionParametrosNum };
export default router;